dev/tools/localization/gen_l10n.dart

```diff
@@ -295,67 +109,190 @@ String generatePluralMethod(Message message) {
+
+  Iterable<LocaleInfo> localesWithCountryCode(String countryCode) {
+    return locales.where((LocaleInfo locale) => locale.countryCode == countryCode);
+  }
+
+  Iterable<LocaleInfo> localesWithScriptCode(String scriptCode) {
+    return locales.where((LocaleInfo locale) => locale.scriptCode == scriptCode);
+  }
+}
+
+String generateLookupBody(Iterable<LocaleInfo> allLocales, String className) {
+  final Map<String, LocalesForLanguage> languageToLocales = <String, LocalesForLanguage>{};
+  for (final LocaleInfo locale in allLocales) {
+    final String languageCode = locale.languageCode;
+    languageToLocales[languageCode] ??= LocalesForLanguage(languageCode);
+    languageToLocales[languageCode].locales.add(locale);
+    if (locale.scriptCode != null)
+      languageToLocales[languageCode].scriptCodes.add(locale.scriptCode);
+    if (locale.countryCode != null)
+      languageToLocales[languageCode].countryCodes.add(locale.countryCode);
+  }
+
+  final Iterable<String> switchClauses = languageToLocales.values.map((LocalesForLanguage localesForLanguage) {
+    if (localesForLanguage.scriptCodes.isEmpty && localesForLanguage.countryCodes.isEmpty) {
+      assert(localesForLanguage.locales.length == 1);
+      final LocaleInfo locale = localesForLanguage.locales[0];
+      return switchClauseTemplate
+        .replaceAll('@(case)', localesForLanguage.languageCode)
+        .replaceAll('@(class)', '$className${locale.camelCase()}');
+    }
+
+    if (localesForLanguage.scriptCodes.isEmpty) {
+      return localesForLanguage.countryCodes.map((String countryCode) {
+        final Iterable<LocaleInfo> locales = localesForLanguage.localesWithCountryCode(countryCode);
+        return countryCodeSwitchTemplate
+          .replaceAll('@(languageCode)', localesForLanguage.languageCode)
+          .replaceAll('@(class)', '$className${LocaleInfo.fromString(localesForLanguage.languageCode).camelCase()}')
+          .replaceAll('@(switchClauses)', locales.map((LocaleInfo locale) {
+            return switchClauseTemplate
+              .replaceAll('@(case)', countryCode)
+              .replaceAll('@(class)', '$className${locale.camelCase()}');
+          }).join('\n'));
+      }).join('\n');
+    }
+    assert(false, 'TBD dispatch on scriptCodes, scriptCodes AND countryCodes');
+    return '';
+  });
+
+  return switchClauses.join('\n    ');
 }
 
-/// The localizations generation class used to generate the localizations
-/// classes, as well as all pertinent Dart files required to internationalize a
-/// Flutter application.
 class LocalizationsGenerator {
   /// Creates an instance of the localizations generator class.
   ///
   /// It takes in a [FileSystem] representation that the class will act upon.
   LocalizationsGenerator(this._fs);
 
-  static RegExp arbFilenameLocaleRE = RegExp(r'^[^_]*_(\w+)\.arb$');
-  static RegExp arbFilenameRE = RegExp(r'(\w+)\.arb$');
-
   final file.FileSystem _fs;
+  // TBD: private data
```

Comment:
Why is this TBD? Does it need to be public for some reason?

Reply:
Oops, I'd meant to make these fields private.